### PR TITLE
[MIG] stock_mts_mto_rule: Migration to 13.0

### PR DESCRIPTION
**Pull request by deivislaya**
Migration Standard

The movements in V13 are now assigned after executing the ``Procurement`` of each ``stock.rule``. 

This is why now, an intermediate step is added to assign the movements of the MTS and release the free quantity to be considered in the MTO. This applies only to partial availability of the product in stock

### CODE REVIEW

**Comment by deivislaya** [comment]
@JulioSerna @randall-vx @hugho-ad 

Please review this PR

Regards!

File: stock_mts_mto_rule/models/stock_rule.py [hunk 1]
```diff
@@ -54,30 +54,51 @@ def get_mto_qty_to_order(self, product, product_qty, product_uom, values):
...
+            else:
+                mts_qty = procurement.product_qty - needed_qty
+                mts_procurement = procurement._replace(product_qty=mts_qty)
+                getattr(self.env["stock.rule"], "_run_%s" % rule.mts_rule_id.action)(
+                    [(mts_procurement, rule.mts_rule_id)]
+                )
+
+                # Search all confirmed stock_moves of mts_procuremet and assign them
+                # to adjust the product's free qty
+                domain = self.env["procurement.group"]._get_moves_to_assign_domain()
```
Comment (rousseldenis) [inline stock_mts_mto_rule/models/stock_rule.py]: Is it possible to make this real multi implementation ? If procurements is real big recordset, you will do a search per record.

So, it's better to build domain before the loop. and assign moves at the end.

Comment (deivislaya) [reply stock_mts_mto_rule/models/stock_rule.py]: Domain applied outside the cycle, however, it is necessary to assign the movements of procuremets_mts before applying the next procurement _mto

**Comment by hugho-ad** [comment]
@rousseldenis 

could you review again please?

**Comment by randall-vx** [comment]
Functionally tested !  :+1: 

1. Create product called "Frutas Secas" (Price: 1500 unit, Cost: 1125, stockeable, MTS & MTO configured in warehouse)
2. Product updated manually to 55 units
3. Update cost manually (just for testing)


image

4. Create SO by 250 units

image

5. Out of stock. Waiting for another operation.

image

 6. PO automatically created by 195 units.

image

cc: @hugho-ad @deivislaya

**Comment by hugho-ad** [comment]
@rousseldenis 

please review!

**Comment by eversatis** [comment]
Is this module avaible for production use now?

**Comment by hugho-ad** [comment]
@eversatis until the merge, but you can use it

**Comment by hugho-ad** [comment]
@moylop260
could you help us with review/merge please

**Review by moylop260 — APPROVED** [review]
👍 

I think the red CI is not relate:

Using just OCA/OCB but not odoo/odoo
```
2020-09-30 15:49:27,272 5720 ERROR openerp_test odoo.addons.stock_vertical_lift_storage_type.tests.test_put: FAIL: TestPut.test_storage_type_put_away
Traceback (most recent call last):
File "/home/travis/build/OCA/stock-logistics-warehouse/stock_vertical_lift_storage_type/tests/test_put.py", line 131, in test_storage_type_put_away
self.assertEqual(move_line.location_dest_id, self.location_1a_x1y1)
AssertionError: stock.location(803,) != stock.location(794,)
```

logger.warning of another module:
```
2020-09-30 17:02:52,650 172 WARNING openerp_test odoo.addons.server_environment.server_env: `running_env` not found. Using default = `test`.
2020-09-30 17:02:52,650 172 WARNING openerp_test odoo.addons.server_environment.server_env: We strongly recommend against using the rc file but instead use an explicit config file or env variable.
```

**Comment by moylop260** [comment]
<s>
@hugho-ad @deivislaya 

Could you help us to finish the following pending:
 - https://github.com/OCA/server-env/pull/57#issuecomment-699029447

In order to get CI green and use ocabot to auto-merge

@pedrobaeza @sbidoul 
Is OCA/OCB error knew it?

</s>

Forget it
It was a outdated build and sha

**Comment by hugho-ad** [comment]
@moylop260 
it's done at https://github.com/OCA/server-env/pull/58 & https://github.com/OCA/server-env/pull/59

**Comment by moylop260** [comment]
So, could you rebase, please?

Maybe it is so old

**Comment by hugho-ad** [comment]
@deivislaya 

could you fix unitests please

**Comment by FaTaToo** [comment]
Hi, I tested this module for a week. It's quite good. Thank you @deivislaya 👍 

However, I found a small bug relating to canceling a confirmed SO.

Suppose you sell 12 units of product A and you already have 5 units in stock. After confirm SO the system will create a new PO to buy more 7 units of product A. Then if you cancel that SO, a warning message will be logged on PO with the wrong number. It shows `12` instead of `7`

image

**Comment by moylop260** [comment]
@deivislaya 

Is possible to add a unittest like @FaTaToo case, please?

**Comment by deivislaya** [comment]
> Hi, I tested this module for a week. It's quite good. Thank you @deivislaya 
> 
> However, I found a small bug relating to canceling a confirmed SO.
> 
> Suppose you sell 12 units of product A and you already have 5 units in stock. After confirm SO the system will create a new PO to buy more 7 units of product A. Then if you cancel that SO, a warning message will be logged on PO with the wrong number. It shows `12` instead of `7`
> 
> image

Hi   @FaTaToo 

It is a feature not related to this module 

More information in

https://github.com/odoo/odoo/blob/13.0/addons/sale_purchase/data/mail_data.xml#L28-L46
https://github.com/odoo/odoo/blob/13.0/addons/sale_purchase/models/sale_order.py#L59

Regards!

cc @moylop260

**Comment by hugho-ad** [comment]
@moylop260  ready to review/merge

red runbot is not related

**Comment by moylop260** [comment]
You are right!

Runbot is making a warning with another module: `stock_location_zone`
because of new way to use queue_job

So, 👍 
Thank you!

**Pull request merged**